### DISCUSSION TITLE
Dependency requirement release 2.18

### DISCUSSION THREAD

**Discussion by elassadi** [avo-hq/avo#1358] [General]
### I am wondering how you guys think about the dependency requirement introduced in the new release 

IMHO dependencies are somehow a limitation, and they should kept as minimal as possible, 
Personally, I don't like the idea to be forced to add gems into my project (again, this is my own personal flavor). 
How about you ? 

https://docs.avohq.io/2.0/upgrade.html#upgrade-from-2-17-to-2-18
``` ruby
# Minimal authorization through OO design and pure Ruby classes
gem "pundit"

# Active Storage makes it simple to upload and reference files
gem "activestorage"

# High-level image processing wrapper for libvips and ImageMagick/GraphicsMagick
gem "image_processing"

# All sorts of useful information about every country packaged as convenient little country objects.
gem "countries"

# Create beautiful JavaScript charts with one line of Ruby
gem "chartkick"
```
### Ideas 

1. Avo should not require any gems, and if there are any dependencies, they should be handled in the gem specs 
for the runtime dependencies 
2. Feature relaying on any dependencies should be optional

The current Release is requiring Pundit and other gems, many rails app out there are using cancancan or 
other Authorization library. 
We should be able to run Avo without any authorization adapter loaded. 

Looking forward to hearing your though about my concern.

cheers 
Moe

**Comment by adrianthedev** [comment]
Hey. Thanks for writing! I really appreciate the feedback.

Maybe I wasn't very clear on why I said that you must include these gems. I'll try to do that now.

We merged two PRs that got us here.

1. https://github.com/avo-hq/avo/pull/1300 is the first change in supporting multiple authorization clients. Meaning that you may use any gem you feel comfortable with (`pundit`, `action_policy`, `cancancan`, etc.). Because of that we pulled `pundit` out from the gemspec (it was required by Avo before this PR) so you can add your own.

Let's say for example you want to use `action_policy`, and not `pundit`. It's a waste to have to pull in `pundit` by default. It's even more wasteful to have Avo require both of them.

That's why we thought that the best strategy is to not require any library by default, but specify to bring your own when needed.

2. https://github.com/avo-hq/avo/pull/1280 is an attempt to slim down the dependencies even more.
For example, very few apps use the `country` field (which internally uses the `countries` gem), but we were bringing in that requirement for every app. 

Same for `chartkick` when you don't use dashboards and `activestorage` if you don't use the file fields.

So the gist is that before, they were getting installed by default whether you'd use them or not, and now we're placing that responsibility in your court in order to slim the dependency tree down.

Does that answer your question?

Do you have any idea on how we could achieve this differently?
